### PR TITLE
Add a null scheduler that helps segmenting away no-op schedules

### PR DESCRIPTION
**Pull request by shmsong**
Remaining task items:

- [x] Add Null scheduler to the infrastructure.
- [x] Compile time matching logic for absolutely null schedules.
- [ ] Runtime matching logic for "dynamically" null schedules, i.e. tensor inputs with 0 size on any dim.
- [ ] Would be helpful to gather as many cases as possible on where a fusion doesn't need any scheduling steps. It looks not easy to come up with a complete list.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/scheduler/all_schedulers.h [hunk 1]
```diff
@@ -10,6 +10,7 @@ namespace cuda {
 
 enum class TORCH_CUDA_CU_API ScheduleHeuristic {
   None,
+  Null,
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler/all_schedulers.h]: I guess `Null` means the scheduler knows to take no action. `None` means the heuristic is yet to be decided.

**Comment by shmsong** [comment]
For dynamically null case, will need to build a structure a root domain map to derive a minimum set of integers to check non-zero, and reject if `zero and not null` on the `canScheduleRuntime` side.

File: torch/csrc/jit/codegen/cuda/scheduler/all_schedulers.h (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/all_schedulers.h]: How about something like path-through instead of null?

**Comment by shmsong** [comment]
@zasdfgbnm looks like either `empty fusion` or `0-sized dimensions` will need some additional handling on the executor side to be able to get lowered and running. Feel free to let me know if there're cases that you'd need along this line.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 2]
```diff
@@ -827,6 +827,103 @@ bool hasNonUniqueBcast(Fusion* fusion) {
...
+ public:
+  explicit NoOpScheduler(
+      Fusion* fusion,
+      SchedulerRuntimeInfo& runtime_info,
+      HeuristicSummary* data_cache = nullptr)
+      : SchedulerEntry(ScheduleHeuristic::NoOp) {
+    params_ = std::make_shared<NoOpHeuristic>();
+  }
+
+  //! Check if the reduction heuristics apply in given fusion
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: nit:
```suggestion
  //! Check if the no-op heuristics apply in given fusion
```

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 3]
```diff
@@ -1485,6 +1589,8 @@ size_t SchedulerEntryHash::operator()(const SchedulerEntry& se) const {
 
 std::string toString(ScheduleHeuristic sh) {
   switch (sh) {
+    case ScheduleHeuristic::NoOp:
+      return "null";
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp]: nit:
```suggestion
      return "no-op";
```

**Review by zasdfgbnm — APPROVED** [review]

**Comment by shmsong** [comment]
Merging for now to unblock.

**Pull request merged**